testsuite/devserver.go

```diff
@@ -0,0 +1,356 @@
+	// Accept nil because all options are "optional".
+	if options == nil {
+		options = &DevServerOptions{}
+	}
+	clientOptions := options.clientOptionsOrDefault()
+
+	exePath, err := downloadIfNeeded(options, clientOptions.Logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if clientOptions.HostPort == "" {
+		// Make sure this is done after downloading to reduce the chance (however slim) that the free port would be used
+		// up by the time the download completes.
+		clientOptions.HostPort, err = getFreeHostPort()
+		if err != nil {
+			return nil, err
+		}
+	}
+	host, port, err := net.SplitHostPort(clientOptions.HostPort)
+	if err != nil {
+		return nil, fmt.Errorf("invalid HostPort: %w", err)
+	}
+
+	args := []string{
+		"server",
+		"start-dev",
+		"--ip", host, "--port", port,
+		"--namespace", clientOptions.Namespace,
+		"--dynamic-config-value", "frontend.enableServerVersionCheck=false",
+		"--dynamic-config-value", "system.enableEagerWorkflowStart=true",
+		"--dynamic-config-value", "system.enableEagerWorkflowStart=true",
+	}
+	if options.LogLevel != "" {
+		args = append(args, "--log-level", options.LogLevel)
+	}
+	if options.LogFormat != "" {
+		args = append(args, "--log-format", options.LogFormat)
+	}
+	cmd := exec.Command(exePath, args...)
+	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
+	clientOptions.Logger.Info("Starting DevServer", "ExePath", exePath, "Args", args)
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed starting: %w", err)
+	}
+
+	returnedClient, err := waitServerReady(ctx, clientOptions)
+	clientOptions.Logger.Info("DevServer ready")
+	if err != nil {
+		return nil, err
+	}
+	return &devServer{client: returnedClient, cmd: cmd}, nil
+}
+
+func downloadIfNeeded(options *DevServerOptions, logger log.Logger) (string, error) {
+	if options.Exe.ExistingPath != "" {
+		return options.Exe.ExistingPath, nil
+	}
+	version := options.Exe.CachedDownload.Version
+	if version == "" {
+		version = "default"
+	}
+	destDir := options.Exe.CachedDownload.DestDir
+	if destDir == "" {
+		destDir = os.TempDir()
+	}
+	var exePath string
+	// Build path based on version and check if already present
+	if version == "default" {
+		exePath = filepath.Join(destDir, "temporal-cli-go-sdk-"+internal.SDKVersion)
+	} else {
+		exePath = filepath.Join(destDir, "temporal-cli-"+version)
+	}
+	if runtime.GOOS == "windows" {
+		exePath += ".exe"
+	}
+	if _, err := os.Stat(exePath); err == nil {
+		return exePath, nil
+	}
+
+	// Build info URL
+	platform := runtime.GOOS
+	if platform != "windows" && platform != "darwin" && platform != "linux" {
+		return "", fmt.Errorf("unsupported platform %v", platform)
+	}
+	arch := runtime.GOARCH
+	if arch != "amd64" && arch != "arm64" {
+		return "", fmt.Errorf("unsupported architecture %v", arch)
+	}
+	infoURL := fmt.Sprintf("https://temporal.download/cli/%v?platform=%v&arch=%v&sdk-name=sdk-go&sdk-version=%v", version, platform, arch, internal.SDKVersion)
```

Comment:
```suggestion
	infoURL := fmt.Sprintf("https://temporal.download/cli/%v?platform=%v&arch=%v&sdk-name=sdk-go&sdk-version=%v", url.QueryEscape(version), platform, arch, internal.SDKVersion)
```

(wasn't as much of a concern for our features repo as it is now)